### PR TITLE
Improve timing extension

### LINKED ISSUE #1098: Timing extension gives misleading information if triggered every n batches
It's due to `every_n_batches` is turned into `after_batch` callback.

The extension should check how many iterations done since the previous call and report average time.

I guess, the same problem exists with `every_n_epochs`.

### PR DESCRIPTION
**Pull request by dmitriy-serdyuk**
- Add prefix argument
- Average timing if it runs every n batches/epochs
- Improve test

Fixes #1098

### CODE REVIEW

File: blocks/extensions/__init__.py [hunk 1]
```diff
@@ -506,23 +506,35 @@ def before_batch(self, batch):
...
     :meth:`~.Algorithm.process_batch` method as well as the time spent
     reading data per batch or epoch. It also reports the time spent
     initializing the algorithm.
 
+    Parameters
+    ----------
+    prefix : str
+        Prefix to be added to the log record, looks the best if ends with
+        ``_``. Defaults to the empty string.
+    \*\*kwargs
```
Comment (dwf) [inline blocks/extensions/__init__.py]: We don't normally document this explicitly, do we?

File: blocks/extensions/__init__.py [hunk 2]
```diff
@@ -506,23 +506,35 @@ def before_batch(self, batch):
...
+        Keyword arguments which are passed to the
+        :class:`.SimpleExtension`.
+
     Notes
     -----
     Add this extension *before* the :class:`Printing` extension.
 
+    Called with callbacks like ``every_n_batches`` this extension this
+    extension averages the time.
```
Comment (dwf) [inline blocks/extensions/__init__.py]: "this extension this extension"

File: blocks/extensions/__init__.py [hunk 3]
```diff
@@ -506,23 +506,35 @@ def before_batch(self, batch):
...
     This adds data about the time spent in the algorithm's
     :meth:`~.Algorithm.process_batch` method as well as the time spent
     reading data per batch or epoch. It also reports the time spent
     initializing the algorithm.
 
+    Parameters
+    ----------
+    prefix : str
+        Prefix to be added to the log record, looks the best if ends with
+        ``_``. Defaults to the empty string.
```
Comment (dwf) [inline blocks/extensions/__init__.py]: Hmm, I think in other places we add the "_" automatically, e.g. DataStreamMonitoring

File: blocks/extensions/__init__.py [hunk 4]
```diff
@@ -549,7 +566,10 @@ def do(self, which_callback, *args):
...
             self.previous[level][action] = self.current[level][action]
             self.current[level][action] = profile['training', 'epoch', action]
-            current_row['time_{}_this_{}'.format(action, level)] = \
-                self.current[level][action] - self.previous[level][action]
-            current_row['time_{}_total'.format(action)] = \
+            this_time = self.prefix + 'time_{}_this_{}'
+            current_row[this_time.format(action, level)] = (
+                (self.current[level][action] - self.previous[level][action]) /
+                self.divisor[level])
+            total_time = self.prefix + 'time_{}_total'
```
Comment (dwf) [inline blocks/extensions/__init__.py]: It seems like we can make weird things happen here if specifying weird things like `every_n_batches=5, after_batch=True`.

Comment (dwf) [inline blocks/extensions/__init__.py]: Actually, how does `level` even get set in the case of `every_n_*` callbacks?

**Comment by dwf** [comment]
Here's an idea for a slightly more robust setup: record the previous batch or epoch for each level, then use the difference between the current one and the previous one as the divisor. This naturally handles the issue I brought up, where if someone has boneheadedly specified `after_batch` and `every_n_batches=5` you will see when the proposed divisor is 0 and can fail or spew warnings appropriately.

**Comment by dmitriy-serdyuk** [comment]
Right, it should be more robust. I also forgot, that a user might create an extension with an arbitrary predicate.

**Comment by dwf** [comment]
Ack, I forgot about predicate-modulated conditions too...

File: blocks/extensions/__init__.py (hunk 4 above)
Comment (dmitriy-serdyuk) [inline blocks/extensions/__init__.py]: Yes, `every_n_*` are triggered as `after_*` callbacks, that was the source of my confusion.

Comment (dmitriy-serdyuk) [inline blocks/extensions/__init__.py]: See here: https://github.com/mila-udem/blocks/blob/master/blocks/extensions/__init__.py#L249

**Comment by dmitriy-serdyuk** [comment]
I changed it in a way you suggested, tests are passing.

File: tests/extensions/test_timing.py [hunk 5]
```diff
@@ -1,8 +1,20 @@
...
+                             extensions=[Timing(prefix='each'),
+                                         Timing(prefix='each_second',
+                                                every_n_epochs=2),
+                                         FinishAfter(after_n_epochs=epochs)])
     main_loop.run()
+    iterations = int(main_loop.log.status['iterations_done'] / epochs)
+    assert_allclose(
+        (main_loop.log[iterations]['each_time_train_this_epoch'] +
+         main_loop.log[iterations]['each_time_train_this_epoch']) / 2,
+        main_loop.log.current_row['each_second_time_train_this_epoch'],
```
Comment (rizar) [inline tests/extensions/test_timing.py]: Our syntax check is currently broken, but I think I see a syntax error here. Can you please run flake8 locally?

Comment (dmitriy-serdyuk) [inline tests/extensions/test_timing.py]: I've run it, formatting is fine. Why do you think it's broken?

Comment (rizar) [inline tests/extensions/test_timing.py]: OK, my bad.

**Comment by rizar** [comment]
I would say LGTM, but @dwf  was the first to start reviewing this PR.

BTW, I will probably add wall-time monitoring to this extension right after this PR is merged.

**Comment by dwf** [comment]
LGTM!

**Pull request merged**